File: backend/app/schemas/checkout.py
import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, EmailStr, Field


class CheckoutItem(BaseModel):
    product_id: uuid.UUID
    variant_id: uuid.UUID
    quantity: int = Field(ge=1, le=99)


class CheckoutAddress(BaseModel):
    first_name: str = Field(min_length=1, max_length=80)
    last_name: str = Field(min_length=1, max_length=80)
    email: EmailStr
    phone: str | None = Field(default=None, max_length=40)
    country: str = Field(min_length=2, max_length=2)
    state: str = Field(min_length=1, max_length=100)
    city: str = Field(min_length=1, max_length=120)
    address_line1: str = Field(min_length=3, max_length=255)
    address_line2: str | None = Field(default=None, max_length=255)
    district: str | None = Field(default=None, max_length=120)
    postal_code: str = Field(min_length=3, max_length=30)
    notes: str | None = Field(default=None, max_length=1000)


class CheckoutCalculateRequest(BaseModel):
    items: list[CheckoutItem] = Field(min_length=1)
    address: CheckoutAddress | None = None
    shipping_method_code: str | None = None
    coupon_code: str | None = None
    currency: str = Field(default="USD", min_length=3, max_length=3)


class CheckoutCreateOrderRequest(CheckoutCalculateRequest):
    address: CheckoutAddress
    idempotency_key: str = Field(min_length=8, max_length=120)


class CheckoutLine(BaseModel):
    product_id: uuid.UUID
    variant_id: uuid.UUID
    product_name: str
    variant_sku: str
    quantity: int
    unit_price: Decimal
    total_price: Decimal
    currency: str


class ShippingQuote(BaseModel):
    code: str
    name: str
    amount: Decimal
    currency: str
    min_days: int
    max_days: int
    tracking_available: bool


class CheckoutTotals(BaseModel):
    subtotal_amount: Decimal
    discount_amount: Decimal
    shipping_amount: Decimal
    tax_amount: Decimal
    total_amount: Decimal
    currency: str
    coupon_code: str | None = None
    shipping_method_code: str | None = None


class CheckoutCalculationResponse(BaseModel):
    items: list[CheckoutLine]
    shipping_methods: list[ShippingQuote]
    totals: CheckoutTotals


class OrderItemRead(CheckoutLine):
    id: uuid.UUID

    class Config:
        from_attributes = True


class OrderRead(BaseModel):
    id: uuid.UUID
    order_number: str
    customer_email: EmailStr
    status: str
    payment_status: str
    fulfillment_status: str
    supplier_status: str
    subtotal_amount: Decimal
    discount_amount: Decimal
    shipping_amount: Decimal
    tax_amount: Decimal
    total_amount: Decimal
    currency: str
    created_at: datetime
    items: list[OrderItemRead] = []

    class Config:
        from_attributes = True

File: frontend/src/types/checkout.ts
export type CheckoutItemPayload = {
  product_id: string;
  variant_id: string;
  quantity: number;
};

export type CheckoutAddress = {
  first_name: string;
  last_name: string;
  email: string;
  phone?: string;
  country: string;
  state: string;
  city: string;
  address_line1: string;
  address_line2?: string;
  district?: string;
  postal_code: string;
  notes?: string;
};

export type CheckoutCalculateRequest = {
  items: CheckoutItemPayload[];
  address?: CheckoutAddress;
  shipping_method_code?: string;
  coupon_code?: string;
  currency: string;
};

export type CheckoutLine = {
  product_id: string;
  variant_id: string;
  product_name: string;
  variant_sku: string;
  quantity: number;
  unit_price: string;
  total_price: string;
  currency: string;
};

export type ShippingQuote = {
  code: string;
  name: string;
  amount: string;
  currency: string;
  min_days: number;
  max_days: number;
  tracking_available: boolean;
};

export type CheckoutTotals = {
  subtotal_amount: string;
  discount_amount: string;
  shipping_amount: string;
  tax_amount: string;
  total_amount: string;
  currency: string;
  coupon_code?: string | null;
  shipping_method_code?: string | null;
};

export type CheckoutCalculation = {
  items: CheckoutLine[];
  shipping_methods: ShippingQuote[];
  totals: CheckoutTotals;
};

export type Order = {
  id: string;
  order_number: string;
  customer_email: string;
  status: string;
  payment_status: string;
  fulfillment_status: string;
  supplier_status: string;
  supplier_order_id?: string | null;
  subtotal_amount: string;
  discount_amount: string;
  shipping_amount: string;
  tax_amount: string;
  total_amount: string;
  currency: string;
  shipping_method_code?: string | null;
  shipping_method_name?: string | null;
  shipping_min_days?: number | null;
  shipping_max_days?: number | null;
  shipping_tracking_available?: boolean;
  created_at: string;
  items: CheckoutLine[];
};

export type PaymentSession = {
  order_number: string;
  gateway: string;
  status: string;
  amount: string;
  currency: string;
  checkout_session_id?: string | null;
  payment_url?: string | null;
  mode: string;
};
